chain-orderer/src/chain-orderer.ts

```diff
@@ -0,0 +1,225 @@
+            if (current_depth > max_depth) {
+                throw new Error(`Max shard search depth (${max_depth}) exceeded`);
+            }
+
+            let current_blocks = await this.dBmtDb.get_shardchain_blocks_by_ids(shardchain_block_ids_to_get, current_mc_block.gen_utime);
+            blocks.push(...current_blocks);
+
+            shardchain_block_ids_to_get = []
+            current_blocks.forEach(b => {
+                if (!root_shardchain_block_ids.has(b.prev_block_id)) {
+                    shardchain_block_ids_to_get.push(b.prev_block_id);
+                }
+
+                if (b.prev_alt_block_id && !root_shardchain_block_ids.has(b.prev_alt_block_id)) {
+                    shardchain_block_ids_to_get.push(b.prev_alt_block_id);
+                }
+            })
+
+            current_depth++;
+        }
+
+        return {
+            master_block: current_mc_block,
+            shard_blocks: blocks,
+        };
+    }
+
+    private async set_chain_order_for_bmt(bmt: Bmt): Promise<void> {
+        if (bmt.master_block.chain_order && 
+            bmt.shard_blocks.reduce((ok, block) => ok && !!block.chain_order, true)) {
+            // we reached the blocks chain-ordered by parser or
+            // the script failed between setting chain order and
+            // updating summary
+            this.verify_bmt(bmt);
+            return;
+        }
+
+        const master_order = toU64String(bmt.master_block.seq_no);
+
+        for (let b_i = 0; b_i < bmt.shard_blocks.length; b_i++) {
+            const block = bmt.shard_blocks[b_i];
+
+            if (block.chain_order) {
+                continue;
+            }
+            
+            const workchain_order = toU64String(block.workchain_id);
+            const seq_no_order = toU64String(block.seq_no);
+            const shard_order = shard_to_reversed_to_U64String(block.shard);
+            const inner_order = workchain_order + seq_no_order + shard_order;
+            const block_order = master_order + inner_order; 
+            
+            await this.set_chain_order_for_block_transactions(block, block_order);
+            await this.dBmtDb.set_chain_order_for_block(block, block_order);
+        }
+
+        const master_block_order = master_order + "m";
+        await this.set_chain_order_for_block_transactions(bmt.master_block, master_block_order);
+        await this.dBmtDb.set_chain_order_for_block(bmt.master_block, master_block_order);
+    }
+
+    private async verify_bmt(bmt: Bmt) {        
+        for (let b_i = 0; b_i < bmt.shard_blocks.length; b_i++) {
+            const block = bmt.shard_blocks[b_i];
+            this.dBmtDb.verify_block_and_transactions(block);
+        }
+
+        this.dBmtDb.verify_block_and_transactions(bmt.master_block);
+    }
+
+    private async set_chain_order_for_block_transactions(block: Block, block_order: string): Promise<void> {
+        const transactions = block.transactions;
```

Comment:
You sort other people data and produce side effect on block's internal data. It is better to get own array copy and sort it instead of original. `const transactions = [...block.transactions];`